examples/src/main/java/alluxio/cli/UnderFileSystemContractTest.java

```diff
@@ -0,0 +1,209 @@
+    if (mFactory == null || !mFactory.supportsPath(mUfsPath)) {
+      LOG.error("%s is not a valid %s path", mUfsPath, mUfsType);
+      System.exit(1);
+    }
+
+    runCommonOperations();
+    if (mUfsType.equals("S3A")) {
+      runS3AOperations();
+    }
+    CliUtils.printPassInfo(true);
+  }
+
+  private void runCommonOperations() throws Exception {
+    InstancedConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+    conf.set(PropertyKey.UNDERFS_LISTING_LENGTH, "50");
+    conf.set(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT, "512B");
+    // Increase the buffer time of journal writes to speed up tests
+    conf.set(PropertyKey.MASTER_JOURNAL_FLUSH_BATCH_TIME_MS, "1sec");
+
+    UnderFileSystem ufs = createUnderFileSystem(conf);
+    // Create the test directory to run tests against
+    String testDir = PathUtils.concatPath(mUfsPath, UUID.randomUUID());
+    UnderFileSystemCommonOperations ops
+        = new UnderFileSystemCommonOperations(mUfsPath, testDir, ufs, conf);
+    try {
+      ops.runTests();
+    } catch (IOException e) {
+      if (mUfsType.equals("S3A")) {
+        List<String> operations = ops.getRelatedS3AOperations();
+        if (operations.size() > 0) {
+          LOG.info("Related S3 operations: "
+              + StringUtils.join(ops.getRelatedS3AOperations(), ","));
+        }
+      }
+      throw e;
+    } finally {
+      cleanup(ufs, testDir);
+    }
+  }
+
+  private void runS3AOperations() throws IOException {
+    InstancedConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+    conf.set(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT, "1MB");
+    conf.set(PropertyKey.MASTER_JOURNAL_FLUSH_BATCH_TIME_MS, "1sec");
+    conf.set(PropertyKey.UNDERFS_S3A_LIST_OBJECTS_VERSION_1, "true");
+    conf.set(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_ENABLED, "true");
+    conf.set(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE, "5MB");
+    conf.set(PropertyKey.UNDERFS_S3A_INTERMEDIATE_UPLOAD_CLEAN_AGE, "0");
+
+    UnderFileSystem ufs = createUnderFileSystem(conf);
+    String testDir = PathUtils.concatPath(mUfsPath, UUID.randomUUID());
+    S3ASpecificOperations ops
+        = new S3ASpecificOperations(mUfsPath, testDir, (S3AUnderFileSystem) ufs, conf);
+    try {
+      ops.runTests();
+    } finally {
+      cleanup(ufs, testDir);
+    }
+  }
+
+  private UnderFileSystem createUnderFileSystem(InstancedConfiguration conf) {
+    UnderFileSystem ufs = mFactory.create(mUfsPath,
+        UnderFileSystemConfiguration.defaults(conf), conf);
+    if (ufs == null) {
+      LOG.error("Failed to create under filesystem");
+      System.exit(1);
+    }
+    return ufs;
+  }
+
+  private void cleanup(UnderFileSystem ufs, String testDir) throws IOException {
+    ufs.deleteDirectory(testDir, DeleteOptions.defaults().setRecursive(true));
+    ufs.close();
+  }
+
+  private void createUnderFileSystemFactory() throws IOException {
+    switch (mUfsType) {
+      case "HDFS":
```

Comment:
instead if enumerating all ufs's, can't we service load from the libs directory?

Reply:
with a dependency of `alluxio-underfs` as `provided` (instead of individual `compiled` ufs) and using `UnderFileSystemFactoryRegistry.find(path, UnderFileSystemConfiguration.defaults())` to load the factory, we are saying that contract tests can only be run from the servers (which have the `lib` and `extensions` directory) instead of bundling up all ufs modules into the client examples jar